Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp

```diff
@@ -34,309 +34,260 @@
+        return true;
     }
-    // Fallthrough.
-    case State::EndBoundary: {
-        if (m_buffer.size() < 2)
-            return false; // Not enough data to check. Return later when there is more data.
-
-        // We'll decide if this is a closing boundary or an opening one.
-        auto* content = m_buffer.data();
-
-        if (content[0] == '-' && content[1] == '-') {
-            // This is a closing boundary. Close down the handler.
-            m_state = State::End;
-            return false;
-        }
-
-        // This was a simple content separator not a closing one.
-        // Go to before the content processing.
-        m_state = State::InBoundary;
-        break;
+    case State::WaitingForHeaderProcessing: {
+        // Wait until completeHeaderProcessing() is called
+        return false;
     }
-    case State::End:
-        // We are done. Nothing to do anymore.
+    case State::WaitingForTerminate: {
+        // Wait until didCompleteMessage() is called
         return false;
-    default:
-        ASSERT_NOT_REACHED();
+    }
+    case State::Terminating: {
+        m_client.didCompleteFromMultipart();
+        m_state = State::End;
         return false;
     }
+    case State::End:
+        return false;
+    }
+
+    return false;
+}
+
+bool CurlMultipartHandle::shouldTerminateIfNeeded()
+{
+    if (m_didCompleteMessage) {
+        m_state = State::Terminating;
+        return true;
+    }
 
-    return true; // There are still things to process, so go for it.
+    return false;
 }
 
-bool CurlMultipartHandle::checkForBoundary(size_t& boundaryStartPosition, size_t& lastPartialMatchPosition)
+std::optional<bool> CurlMultipartHandle::findBoundary(size_t& matchPos, size_t& endPos, bool hasCloseDelimiter)
 {
     auto contentLength = m_buffer.size();
+    const auto content = m_buffer.data();
+    auto boundaryLength = m_boundary.length();
+    const auto boundary = m_boundary.latin1().data();
 
-    boundaryStartPosition = notFound;
-    lastPartialMatchPosition = contentLength;
+    matchPos = notFound;
+    endPos = notFound;
+    hasCloseDelimiter = false;
 
-    if (contentLength < m_boundary.length())
-        return false;
+    for (size_t pos = 0; pos < contentLength; ++pos) {
+        size_t cmpLength = boundaryLength <= contentLength - pos ? boundaryLength : contentLength - pos;
 
-    const auto content = m_buffer.data();
+        if (memcmp(content + pos, boundary, cmpLength))
+            continue;
 
-    for (size_t i = 0; i < contentLength - m_boundary.length(); ++i) {
-        auto length = matchedLength(content + i);
-        if (length == m_boundary.length()) {
-            boundaryStartPosition = i;
+        auto afterBoundaryPos = pos + boundaryLength;
+
+        // The initial CRLF is considered to be attached to the boundary delimiter line rather than
+        // part of the preceding part. (See RFC2046 [5.1.1. Common Syntax])
+        if (pos - 2 >= 0 && !memcmp(content + pos - 2, "\r\n", 2))
```

Comment:
Why not `pos >=2` here. IIRC, `unsigned - signed` is unsigned. `pos - 2 >= 0` is always true.

Reply:
Fixed